### PR TITLE
GEODE-9203: Implement Radish PUBSUB NUMSUB subcommand

### PR DESCRIPTION
**Pull request by jdeppe-pivotal**
Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/pubsub/AbstractSubCommandsIntegrationTest.java [hunk 1]
```diff
@@ -41,24 +44,32 @@
...
   }
 
   @After
   public void teardown() {
     if (mockSubscriber.getSubscribedChannels() > 0) {
       mockSubscriber.unsubscribe();
     }
+    if (mockSubscriber.getSubscribedChannels() > 0) {
+      mockSubscriber.punsubscribe();
+    }
```
Comment (DonalEvans) [inline geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/pubsub/AbstractSubCommandsIntegrationTest.java:72]: Can this be combined with the if statement above it?

File: geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/pubsub/AbstractSubCommandsIntegrationTest.java [hunk 2]
```diff
@@ -191,6 +201,67 @@ public void channels_shouldNotReturnDuplicates_givenMultipleSubscribersToSameCha
...
+  public void numsub_shouldReturnListOfChannelsWithSubscriberCount_whenCalledWithActiveChannels() {
+    Jedis subscriber2 = new Jedis(BIND_ADDRESS, getPort(), REDIS_CLIENT_TIMEOUT);
+    MockSubscriber fooAndBarSubscriber = new MockSubscriber();
+
+    executor.submit(() -> subscriber.subscribe(mockSubscriber, "foo"));
+    executor.submit(() -> subscriber2.subscribe(fooAndBarSubscriber, "foo", "bar"));
+    waitFor(() -> mockSubscriber.getSubscribedChannels() == 1
+        && fooAndBarSubscriber.getSubscribedChannels() == 2);
+
+    HashMap<String, String> result =
```
Comment (DonalEvans) [inline geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/pubsub/AbstractSubCommandsIntegrationTest.java:227]: If you use `Map<String, String>` here, you can remove the cast.

File: geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/pubsub/AbstractSubCommandsIntegrationTest.java [hunk 3]
```diff
@@ -191,6 +201,67 @@ public void channels_shouldNotReturnDuplicates_givenMultipleSubscribersToSameCha
...
+    waitFor(() -> fooAndBarSubscriber.getSubscribedChannels() == 0);
+    subscriber2.close();
+  }
+
+  @Test
+  public void numsub_shouldReturnChannelNameWithZero_whenCalledWithChannelWithNoSubscribers() {
+    executor.submit(() -> subscriber.subscribe(mockSubscriber, "foo"));
+    waitFor(() -> mockSubscriber.getSubscribedChannels() == 1);
+
+    HashMap<String, String> result =
```
Comment (DonalEvans) [inline geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/pubsub/AbstractSubCommandsIntegrationTest.java:246]: If you use `Map<String, String>` here, you can remove the cast.

File: geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/pubsub/AbstractSubCommandsIntegrationTest.java [hunk 4]
```diff
@@ -191,6 +201,67 @@ public void channels_shouldNotReturnDuplicates_givenMultipleSubscribersToSameCha
...
+    assertThat(result.containsKey("bar")).isTrue();
+    assertThat(result.get("bar")).isEqualTo("0");
+  }
+
+  @Test
+  public void numsub_shouldReturnPatternWithZero_whenCalledWithPatternWithNoChannelSubscribers() {
+    executor.submit(() -> subscriber.psubscribe(mockSubscriber, "f*"));
+    waitFor(() -> mockSubscriber.getSubscribedChannels() == 1);
+
+    HashMap<String, String> result = (HashMap<String, String>) introspector.pubsubNumSub("f*");
```
Comment (DonalEvans) [inline geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/pubsub/AbstractSubCommandsIntegrationTest.java:258]: If you use `Map<String, String>` here, you can remove the cast.

File: geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/executor/pubsub/PubSubExecutor.java [hunk 5]
```diff
@@ -33,28 +32,42 @@
...
+
+        List<byte[]> channelsResponse = doChannels(processedCommand, context);
+        return RedisResponse.array(channelsResponse);
 
-    // in a subsequent story, a new parameter requirement class
-    // specific to subCommands might be a better way to do this
-    if (command.getProcessedCommand().size() > 3) {
-      return RedisResponse
-          .error(String.format(ERROR_UNKNOWN_PUBSUB_SUBCOMMAND, new String(subCommand)));
+      case "numsub":
```
Comment (DonalEvans) [inline geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/executor/pubsub/PubSubExecutor.java:47]: After rebasing on develop, please add
```
public static final byte[] bNUMSUB = stringToBytes("NUMSUB");
```
to the "PubSubExecutor" section in `StringBytesGlossary` and use the `Coder.equalsIgnoreCaseBytes()` method here.

File: geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/pubsub/AbstractSubCommandsIntegrationTest.java [hunk 6]
```diff
@@ -191,6 +201,67 @@ public void channels_shouldNotReturnDuplicates_givenMultipleSubscribersToSameCha
...
+
+    HashMap<String, String> result =
+        (HashMap<String, String>) introspector.pubsubNumSub("bar");
+
+    assertThat(result.containsKey("bar")).isTrue();
+    assertThat(result.get("bar")).isEqualTo("0");
+  }
+
+  @Test
+  public void numsub_shouldReturnPatternWithZero_whenCalledWithPatternWithNoChannelSubscribers() {
```
Comment (DonalEvans) [inline geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/pubsub/AbstractSubCommandsIntegrationTest.java:254]: Could we also get a test for the behaviour when numsub is used with a pattern that does have channel subscribers?

**Review by DonalEvans — CHANGES_REQUESTED** [review]
Some minor clean-up and changes associated with rebasing.

**Review by sabbey37 — COMMENTED** [review]
This looks good once Donal's changes are addressed!  I haven't had a chance to play around with it on the cli, but will let you know if anything comes up. Could we also enable relevant Redis tcl tests for NUMSUB and CHANNELS as a part of this story (it will require generating a new patch file).

File: geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/pubsub/AbstractSubCommandsIntegrationTest.java (hunk 6 above)
Comment (jdeppe-pivotal) [reply geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/pubsub/AbstractSubCommandsIntegrationTest.java]: I think I added what you were asking.

File: geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/executor/pubsub/PubSubExecutor.java (hunk 5 above)
Comment (jdeppe-pivotal) [reply geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/executor/pubsub/PubSubExecutor.java]: Done

File: geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/pubsub/AbstractSubCommandsIntegrationTest.java (hunk 1 above)
Comment (jdeppe-pivotal) [reply geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/pubsub/AbstractSubCommandsIntegrationTest.java:72]: There is a race condition here. In the jedis code,  if the subscribers end up going to 0 then there is a possiblilty of a NPE for subsequent (p)unsubscribe calls if we don't first do the check.

File: geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/pubsub/AbstractSubCommandsIntegrationTest.java (hunk 2 above)
Comment (jdeppe-pivotal) [reply geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/pubsub/AbstractSubCommandsIntegrationTest.java]: Done

File: geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/pubsub/AbstractSubCommandsIntegrationTest.java (hunk 3 above)
Comment (jdeppe-pivotal) [reply geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/pubsub/AbstractSubCommandsIntegrationTest.java]: Done

File: geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/pubsub/AbstractSubCommandsIntegrationTest.java (hunk 4 above)
Comment (jdeppe-pivotal) [reply geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/pubsub/AbstractSubCommandsIntegrationTest.java]: Done

File: geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/pubsub/AbstractSubCommandsIntegrationTest.java (hunk 6 above)
Comment (DonalEvans) [reply geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/pubsub/AbstractSubCommandsIntegrationTest.java]: Yep, looks good, thanks.

**Review by DonalEvans — APPROVED** [review]

**Review by sabbey37 — APPROVED** [review]
Thank you for enabling those tcl tests!

**Pull request merged**